src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -131,11 +131,49 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
                 var conversions = new TypeConversions(corLibrary);
                 var location = singleDeclaration.NameLocation;
 
-                foreach (var @interface in interfaces)
+                foreach (var baseInterface in interfaces)
                 {
-                    @interface.CheckAllConstraints(conversions, location, diagnostics);
+                    baseInterface.CheckAllConstraints(conversions, location, diagnostics);
+                }
+
+                // we check direct base interfaces that have the InternalImplementationOnly attribute,
+                // giving an error when such an attributed interface appears in the interface list.
+                foreach (var baseInterface in this.InterfacesNoUseSiteDiagnostics(null))
+                {
+                    EnsureExtensible(baseInterface, location, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Produce an error if the given base interface may not be extended by this type due to the
+        /// presence of the attribute System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private void EnsureExtensible(NamedTypeSymbol baseInterface, SourceLocation location, DiagnosticBag diagnostics)
+        {
+            if (this.ContainingAssembly != baseInterface.ContainingAssembly &&
+                HasInternalImplementationOnlyAttribute(baseInterface) &&
+                !this.ContainingAssembly.HasInternalAccessTo(baseInterface.ContainingAssembly))
+            {
+                diagnostics.Add(ErrorCode.ERR_InternalImplementationOnly, location, this, baseInterface);
+            }
+        }
+
+        /// <summary>
+        /// Return true of the given interface type contains an attribute of the
+        /// name System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private static bool HasInternalImplementationOnlyAttribute(NamedTypeSymbol baseInterface)
+        {
+            foreach (var att in baseInterface.GetAttributes())
+            {
+                if (att.IsTargetAttribute("System.Runtime.CompilerServices", "InternalImplementationOnlyAttribute"))
```

Comment:
This is a very inefficient way of checking for presence of a well-known attribute. In compiler, we should avoid realizing attributes on metadata and retargeting symbols. Source symbols can benefit from caching of the value too. I suggest doing the following:
- Follow the pattern of NamedTypeSymbol.HasSpecialName property. It is declared abstract and overridden as appropriate by derived classes. 
- Add attribute description to AttributeDescription. Detect and cache presence of the attribute on source symbols in SourceNamedTypeSymbol.DecodeWellKnownAttribute, follow SpecialNameAttribute pattern.
- For PENamedTypeSymbol, delegate check to PEModule, follow the pattern of PEModule.HasRequiredAttributeAttribute. Consider caching the result on the symbol if it can be done without significantly bloating it.